Panic on visor v0.7.1

### Describe the bug:

Running visor v0.7.1 with all tasks enabled on a watch job with confidence: 5 exporting to DB storage. I was seeing this occur frequently and assumed memory pressure had been causing this. After migrating this to its own instance (to avoid contending with other deployments), I caught a full trace in the logs:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x2345868]

goroutine 244903 [running]:
go.opentelemetry.io/otel/sdk/trace.(*evictedQueue).add(...)
/go/pkg/mod/go.opentelemetry.io/otel/sdk@v0.12.0/trace/evictedqueue.go:33
go.opentelemetry.io/otel/sdk/trace.(*span).addEventWithTimestamp(0xc0d8700230, 0xc02f1d3bb7bc6af1, 0x43a7f27e38, 0x6344840, 0x3ab1e8a, 0x5, 0xc0907eec00, 0x2, 0x2)
/go/pkg/mod/go.opentelemetry.io/otel/sdk@v0.12.0/trace/span.go:217 +0x148
go.opentelemetry.io/otel/sdk/trace.(*span).End(0xc0d8700230, 0x0, 0x0, 0x0)
/go/pkg/mod/go.opentelemetry.io/otel/sdk@v0.12.0/trace/span.go:126 +0x565
panic(0x366e800, 0x6181890)
/usr/local/go/src/runtime/panic.go:971 +0x499
panic(0x366e800, 0x6181890)
/usr/local/go/src/runtime/panic.go:965 +0x1b9
go.opentelemetry.io/otel/sdk/trace.(*evictedQueue).add(...)
/go/pkg/mod/go.opentelemetry.io/otel/sdk@v0.12.0/trace/evictedqueue.go:33
go.opentelemetry.io/otel/sdk/trace.(*span).addEventWithTimestamp(0xc0cb52e620, 0xc02f1d3bb7bc1270, 0x43a7f225c0, 0x6344840, 0x3ab1e8a, 0x5, 0xc0907eeb80, 0x2, 0x2)
/go/pkg/mod/go.opentelemetry.io/otel/sdk@v0.12.0/trace/span.go:217 +0x148
go.opentelemetry.io/otel/sdk/trace.(*span).End(0xc0cb52e620, 0x0, 0x0, 0x0)
/go/pkg/mod/go.opentelemetry.io/otel/sdk@v0.12.0/trace/span.go:126 +0x565
panic(0x366e800, 0x6181890)
/usr/local/go/src/runtime/panic.go:965 +0x1b9
github.com/filecoin-project/sentinel-visor/tasks/actorstate.ExtractMinerPoSts.func2(0xc0e5093ab0, 0xc0b8bbe120, 0x4)
/go/src/github.com/filecoin-project/sentinel-visor/tasks/actorstate/miner.go:479 +0x593
github.com/filecoin-project/sentinel-visor/tasks/actorstate.ExtractMinerPoSts(0x48a2f20, 0xc2c6d90f00, 0xc0f23b3d30, 0xc0eb1723c0, 0x7f1c0c854180, 0xc002288000, 0x0, 0x0, 0x0, 0x0, ...)
/go/src/github.com/filecoin-project/sentinel-visor/tasks/actorstate/miner.go:509 +0x4a8
github.com/filecoin-project/sentinel-visor/tasks/actorstate.StorageMinerExtractor.Extract(0x48a2f20, 0xc8522dda10, 0xc0f0ae4948, 0x16, 0xc0d9b1e2a0, 0x26, 0x0, 0xc1b0f364e0, 0xc0b8bbe120, 0x4, ...)
/go/src/github.com/filecoin-project/sentinel-visor/tasks/actorstate/miner.go:74 +0x9b3
github.com/filecoin-project/sentinel-visor/tasks/actorstate.(*Task).runActorStateExtraction(0xc0bde7cf40, 0x48a2f20, 0xc8522dd9e0, 0xc0d8eef140, 0xc121f4dc00, 0xc0df1e4e18, 0x7, 0xc0f0ae4948, 0x16, 0xc0d9b1e2a0, ...)
/go/src/github.com/filecoin-project/sentinel-visor/tasks/actorstate/task.go:179 +0x4ce
created by github.com/filecoin-project/sentinel-visor/tasks/actorstate.(*Task).ProcessActors
/go/src/github.com/filecoin-project/sentinel-visor/tasks/actorstate/task.go:94 +0x8d5
```
### Steps to Reproduce:
EDIT: panics have not occurred June 30 @ ~1:20am UTC
Intermittently reproducible, but seems to eventually occur on a regular interval in our production EKS cluster. (See @placer14 for details.)

### Visor Version: docker image `filecoin/sentinel-visor:v0.7.1` on mainnet staging

--- a/chain/indexer.go
+++ b/chain/indexer.go
@@ -279,11 +279,14 @@ func (t *TipSetIndexer) TipSet(ctx context.Context, ts *types.TipSet) error {
 				log.Errorw("child tipset is not on the same chain as parent", "height", ts.Height(), "child", child.Key(), "parent", parent.Key())
 
 				// We need to report that all message and actor tasks were skipped
+				reason := "child tipset was not on the same chain as parent"
 				for name := range t.messageProcessors {
-					taskOutputs[name] = model.PersistableList{t.buildSkippedTipsetReport(ts, name, start, "child tipset was not on the same chain as parent")}
+					taskOutputs[name] = model.PersistableList{t.buildSkippedTipsetReport(ts, name, start, reason)}
+					ll.Infow("task skipped", "task", name, "reason", reason)
 				}
 				for name := range t.actorProcessors {
-					taskOutputs[name] = model.PersistableList{t.buildSkippedTipsetReport(ts, name, start, "child tipset was not on the same chain as parent")}
+					taskOutputs[name] = model.PersistableList{t.buildSkippedTipsetReport(ts, name, start, reason)}
+					ll.Infow("task skipped", "task", name, "reason", reason)
 				}
 			}
 		}

--- a/chain/indexer.go
+++ b/chain/indexer.go
@@ -201,76 +201,89 @@ func (t *TipSetIndexer) TipSet(ctx context.Context, ts *types.TipSet) error {
 				t.closer = closer
 			}
 
-			// If we have message processors then extract the messages and receipts
-			if len(t.messageProcessors) > 0 {
-				tsMsgs, err := t.node.GetExecutedAndBlockMessagesForTipset(ctx, child, parent)
-				if err == nil {
-					// Start all the message processors
-					for name, p := range t.messageProcessors {
-						inFlight++
-						go t.runMessageProcessor(tctx, p, name, child, parent, tsMsgs.Executed, tsMsgs.Block, results)
-					}
-				} else {
-					ll.Errorw("failed to extract messages", "error", err)
-					terr := xerrors.Errorf("failed to extract messages: %w", err)
-					// We need to report that all message tasks failed
-					for name := range t.messageProcessors {
-						report := &visormodel.ProcessingReport{
-							Height:         int64(ts.Height()),
-							StateRoot:      ts.ParentState().String(),
-							Reporter:       t.name,
-							Task:           name,
-							StartedAt:      start,
-							CompletedAt:    time.Now(),
-							Status:         visormodel.ProcessingStatusError,
-							ErrorsDetected: terr,
+			if types.CidArrsEqual(child.Parents().Cids(), parent.Cids()) {
+				// If we have message processors then extract the messages and receipts
+				if len(t.messageProcessors) > 0 {
+					tsMsgs, err := t.node.GetExecutedAndBlockMessagesForTipset(ctx, child, parent)
+					if err == nil {
+						// Start all the message processors
+						for name, p := range t.messageProcessors {
+							inFlight++
+							go t.runMessageProcessor(tctx, p, name, child, parent, tsMsgs.Executed, tsMsgs.Block, results)
 						}
-						taskOutputs[name] = model.PersistableList{report}
+					} else {
+						ll.Errorw("failed to extract messages", "error", err)
+						terr := xerrors.Errorf("failed to extract messages: %w", err)
+						// We need to report that all message tasks failed
+						for name := range t.messageProcessors {
+							report := &visormodel.ProcessingReport{
+								Height:         int64(ts.Height()),
+								StateRoot:      ts.ParentState().String(),
+								Reporter:       t.name,
+								Task:           name,
+								StartedAt:      start,
+								CompletedAt:    time.Now(),
+								Status:         visormodel.ProcessingStatusError,
+								ErrorsDetected: terr,
+							}
+							taskOutputs[name] = model.PersistableList{report}
+						}
+
 					}
 				}
-			}
 
-			// If we have actor processors then find actors that have changed state
-			if len(t.actorProcessors) > 0 {
-				changesStart := time.Now()
-				var err error
-				var changes map[string]types.Actor
-				// special case, we want to extract all actor states from the genesis block.
-				if parent.Height() == 0 {
-					changes, err = t.getGenesisActors(ctx)
-				} else {
-					changes, err = t.stateChangedActors(tctx, parent.ParentState(), child.ParentState())
-				}
-				if err == nil {
-					ll.Debugw("found actor state changes", "count", len(changes), "time", time.Since(changesStart))
-					if t.addressFilter != nil {
-						for addr := range changes {
-							if !t.addressFilter.Allow(addr) {
-								delete(changes, addr)
+				// If we have actor processors then find actors that have changed state
+				if len(t.actorProcessors) > 0 {
+					changesStart := time.Now()
+					var err error
+					var changes map[string]types.Actor
+					// special case, we want to extract all actor states from the genesis block.
+					if parent.Height() == 0 {
+						changes, err = t.getGenesisActors(ctx)
+					} else {
+						changes, err = t.stateChangedActors(tctx, parent.ParentState(), child.ParentState())
+					}
+					if err == nil {
+						ll.Debugw("found actor state changes", "count", len(changes), "time", time.Since(changesStart))
+						if t.addressFilter != nil {
+							for addr := range changes {
+								if !t.addressFilter.Allow(addr) {
+									delete(changes, addr)
+								}
 							}
 						}
-					}
-					for name, p := range t.actorProcessors {
-						inFlight++
-						go t.runActorProcessor(tctx, p, name, child, parent, changes, results)
-					}
-				} else {
-					ll.Errorw("failed to extract actor changes", "error", err)
-					terr := xerrors.Errorf("failed to extract actor changes: %w", err)
-					// We need to report that all actor tasks failed
-					for name := range t.actorProcessors {
-						report := &visormodel.ProcessingReport{
-							Height:         int64(ts.Height()),
-							StateRoot:      ts.ParentState().String(),
-							Reporter:       t.name,
-							Task:           name,
-							StartedAt:      start,
-							CompletedAt:    time.Now(),
-							Status:         visormodel.ProcessingStatusError,
-							ErrorsDetected: terr,
+						for name, p := range t.actorProcessors {
+							inFlight++
+							go t.runActorProcessor(tctx, p, name, child, parent, changes, results)
 						}
-						taskOutputs[name] = model.PersistableList{report}
+					} else {
+						ll.Errorw("failed to extract actor changes", "error", err)
+						terr := xerrors.Errorf("failed to extract actor changes: %w", err)
+						// We need to report that all actor tasks failed
+						for name := range t.actorProcessors {
+							report := &visormodel.ProcessingReport{
+								Height:         int64(ts.Height()),
+								StateRoot:      ts.ParentState().String(),
+								Reporter:       t.name,
+								Task:           name,
+								StartedAt:      start,
+								CompletedAt:    time.Now(),
+								Status:         visormodel.ProcessingStatusError,
+								ErrorsDetected: terr,
+							}
+							taskOutputs[name] = model.PersistableList{report}
+						}
 					}
+				}
+			} else {
+				log.Errorw("child tipset is not on the same chain as parent", "height", ts.Height(), "child", child.Key(), "parent", parent.Key())
+
+				// We need to report that all message and actor tasks were skipped
+				for name := range t.messageProcessors {
+					taskOutputs[name] = model.PersistableList{t.buildSkippedTipsetReport(ts, name, start, "child tipset was not on the same chain as parent")}
+				}
+				for name := range t.actorProcessors {
+					taskOutputs[name] = model.PersistableList{t.buildSkippedTipsetReport(ts, name, start, "child tipset was not on the same chain as parent")}
 				}
 			}
 		}

--- a/chain/indexer.go
+++ b/chain/indexer.go
@@ -276,10 +276,13 @@ func (t *TipSetIndexer) TipSet(ctx context.Context, ts *types.TipSet) error {
 					}
 				}
 			} else {
-				log.Errorw("child tipset is not on the same chain as parent", "height", ts.Height(), "child", child.Key(), "parent", parent.Key())
+				// TODO: we could fetch the parent stateroot and proceed to index this tipset. However this will be
+				// slower and increases the likelihood that we exceed the processing window and cause the next
+				// tipset to be skipped completely.
+				log.Errorw("mismatching child and parent tipsets", "height", ts.Height(), "child", child.Key(), "parent", parent.Key())
 
 				// We need to report that all message and actor tasks were skipped
-				reason := "child tipset was not on the same chain as parent"
+				reason := "tipset did not have expected parent or child"
 				for name := range t.messageProcessors {
 					taskOutputs[name] = model.PersistableList{t.buildSkippedTipsetReport(ts, name, start, reason)}
 					ll.Infow("task skipped", "task", name, "reason", reason)

